### PR TITLE
Implementation for freezeLeftCount; various refactors; upgrade react …

### PR DESCRIPTION
**Pull request by paulbrom**
This is an implementation for a new **leftFreezeCount** property which will freeze the lefthand columns of the table while allowing you to scroll the righthand columns.  The handleScroll implementation was adjusted to help support this and the default mouse wheel was disabled to ensure we can scroll the tables in tandem on scroll.

Some refactoring was done to break up bulk lodash and use class properties, as well as upgrade the linter and confirm to stricter lint standards.   This package makes agressive use of findDOMNode and such usages were not eliminated (nor may it be possible to do so...)

### CODE REVIEW

File: src/utils/keyHandler.jsx [hunk 1]
```diff
@@ -30,24 +36,34 @@ class KeyHandler extends Component {
...
-      const refNode = ReactDOM.findDOMNode(getInputRef()); // eslint-disable-line max-len, react/no-find-dom-node
-      let node = evt.target;
-      let isDescendant = false;
-      do {
-        if (node === refNode) {
-          isDescendant = true;
+      let inputRef = getInputRef();
+      if (!_isArray(inputRef)) {
+        inputRef = [inputRef];
+      }
```
Comment (paulbrom) [inline src/utils/keyHandler.jsx:45]: I want keys bubbling up to either table (`cellGrid` or `cellGridFreeze`) to be handled seamlessly bu this component, so I added tuspport for `getInputRef` to return an array

File: src/table/virtualizedDataTable.jsx [hunk 2]
```diff
@@ -22,42 +26,375 @@ const DEFAULT_HIGHLIGHT_ROW_COLOR = '#B3E5FC';
...
+  'Space',
+  'Tab',
+];
+for (let keyOn = 0; keyOn < 10; keyOn += 1) {
+  KEYS_TO_SNIFF.push(`Digit${keyOn}`);
+}
+for (let keyOn = 48; keyOn < 91; keyOn += 1) {
+  KEYS_TO_SNIFF.push(`Key${String.fromCharCode(keyOn)}`);
+}
+
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:58]: Next time I need to make changes here I will switch to styled components and add support for them as well

File: src/table/virtualizedDataTable.jsx [hunk 3]
```diff
@@ -140,48 +498,90 @@ class VirtualizedDataTable extends Component {
...
+  prvHandleWheel = (evt) => {
+    if (this.prvCellGridDOM) {
+      evt.preventDefault();
+
+      const { wheelDelta } = this.props;
+      const { state: { scrollTop, scrollLeft } } = this.prvCellGridRef;
+      const delta = evt.deltaY || evt.detail || evt.wheelDelta;
+      const newScrollTop = Math.max(0, scrollTop + ((delta > 0) ? wheelDelta : -wheelDelta));
+      this.prvCellGridDOM.scrollTop = newScrollTop;
+
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:529]: Force a scroll event on wheel, which will be handled by our scroll handler.   Disallow automatic handling of the mouse wheel in favor of always using the scroll handler, which will keep the tables in unison

File: src/table/virtualizedDataTable.jsx [hunk 4]
```diff
@@ -140,48 +498,90 @@ class VirtualizedDataTable extends Component {
...
-      const groupHeader = ReactDOM.findDOMNode(this.prvGroupHeader); // eslint-disable-line max-len, react/no-find-dom-node
-      if (groupHeader) {
-        groupHeader.style.transform = transformStyle;
-      }
-      const headerGrid = ReactDOM.findDOMNode(this.prvHeaderGrid); // eslint-disable-line max-len, react/no-find-dom-node
-      if (headerGrid) {
-        headerGrid.style.transform = transformStyle;
-      } else if (this.prvHeaderGrid && this.prvHeaderGrid._scrollingContainer) { // eslint-disable-line max-len, no-underscore-dangle
-        this.prvHeaderGrid._scrollingContainer.style.transform = transformStyle; // eslint-disable-line max-len, no-underscore-dangle
-      }
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:176]: `handleLeft` now holds the legacy behavior, which kept the header table in sync with the grid table

File: src/table/virtualizedDataTable.jsx [hunk 5]
```diff
@@ -140,48 +498,90 @@ class VirtualizedDataTable extends Component {
...
+          groupHeader.style.transform = transformStyle;
+        }
+        const headerGrid = ReactDOM.findDOMNode(this.prvHeaderGridRef); // eslint-disable-line max-len, react/no-find-dom-node
+        if (headerGrid) {
+          headerGrid.style.transform = transformStyle;
+        } else if (this.prvHeaderGridRef && this.prvHeaderGridRef._scrollingContainer) { // eslint-disable-line max-len, no-underscore-dangle
+          this.prvHeaderGridRef._scrollingContainer.style.transform = transformStyle; // eslint-disable-line max-len, no-underscore-dangle
+        }
+      };
+
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:557]: `handleTop` is the new behavior which keeps the non-frozen grid and its freeze table in sync

File: src/table/virtualizedDataTable.jsx [hunk 6]
```diff
@@ -190,31 +590,29 @@ class VirtualizedDataTable extends Component {
...
         allowRowSelect,
         allowRangeSelect,
         allowMultiSelect,
         onSelectionChange,
         onCellClick,
-        rowCount,
-        rowsCount,
       } = this.props;
-      const rowCountUse = _.isNumber(rowCount) ? rowCount : rowsCount;
+      const { rowCount } = this.state;
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:611]: The above computation is now being done in `getDerivedStateFromProps`

File: src/table/virtualizedDataTable.jsx [hunk 7]
```diff
@@ -932,242 +1329,153 @@ class VirtualizedDataTable extends Component {
         [COLUMN_END]: forPaste ? -1 : focusedCell.columnIndex,
       }])
     );
-  }
+  };
 
   /* ------ END Selection Range Management Methods ------ */
 
   /* ------ Rendering Methods ------ */
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1337]: This was moved to top of file.  No changes except for adding `overflow: initial` to headerTable, and `display: flex` to headerDiv and tableDiv

File: src/table/virtualizedDataTable.jsx [hunk 8]
```diff
@@ -932,242 +1329,153 @@ class VirtualizedDataTable extends Component {
...
-      },
+  prvRecomputeTables = () => {
+    const recomputeTable = (grid, forceOverflowInitial) => {
+      if (grid) {
+        grid.recomputeGridSize();
+        if (forceOverflowInitial) {
+          // eslint-disable-next-line react/no-find-dom-node
+          const domGrid = ReactDOM.findDOMNode(grid);
+          if (domGrid) {
+            domGrid.style.overflow = 'initial';
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1346]: HACK:  This styling is needed to remove scrollbars in the frozen part of the table and ensure proper layout.   Unfortunately, react-virtualized reapplies this, so we need to reapply it also.  Ugly, but there is no great solution without PR against react-virtualized.

File: src/table/virtualizedDataTable.jsx [hunk 9]
```diff
@@ -932,242 +1329,153 @@ class VirtualizedDataTable extends Component {
...
         <div
-          style={_.assign({}, styles.headerCell, {
+          style={{
+            ...STYLES.headerCell,
             height: cellHeight,
-          })}
+          }}
         >
-          {resizer}
-          {clonedCell}
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1114]: `tallDragLine`, `resizer`, and `clonedCell` all inlined below

File: src/table/virtualizedDataTable.jsx [hunk 10]
```diff
@@ -1191,214 +1499,89 @@ class VirtualizedDataTable extends Component {
...
       columnKey,
       cellRenderer,
       headerRenderer,
       flexShrink: 0,
-      style: styles.column,
+      style: STYLES.column,
       width: columnWidth,
-    });
-  }
-
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1203]: This function was inlined below, in the spirit of removing render subfunctions as much as possible

File: src/table/virtualizedDataTable.jsx [hunk 11]
```diff
@@ -1191,214 +1499,89 @@ class VirtualizedDataTable extends Component {
...
-            width: child.props.width || 0,
-          });
-        } else {
-          throw new Error('unexpected child - only Columns or ColumnGroups can be children of VirtualizedDataTable!');
-        }
-      }
-    });
+    } = this.state;
+    const cellGridHeight = height - (headerHeight + 2); // subtract 2 because of borders
+    const columns = columnInfo.map(columnInfoCur => this.prvRenderTableColumn(columnInfoCur));
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1552]: Above we take the set of columns computed in `getDerivedStateFromProps` (formerly right here in `render` as seen above) and render them with `prvRenderTableColumn`, which was a render function and also required access to the component instance and couldn't be done from a static method

File: src/table/virtualizedDataTable.jsx [hunk 12]
```diff
@@ -1460,101 +1641,161 @@ class VirtualizedDataTable extends Component {
       }
     }
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1643]: This was moved to top of file

File: src/table/virtualizedDataTable.jsx [hunk 13]
```diff
@@ -1460,101 +1641,161 @@ class VirtualizedDataTable extends Component {
...
-      'Semicolon',
-      'Slash',
-      'Space',
-      'Tab',
-    ];
-    for (let keyOn = 0; keyOn < 10; keyOn += 1) {
-      keysToSniff.push(`Digit${keyOn}`);
-    }
-    for (let keyOn = 48; keyOn < 91; keyOn += 1) {
-      keysToSniff.push(`Key${String.fromCharCode(keyOn)}`);
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1490]: Determine if we need to apply `scrollToColumn` to snap a column into view (like when walking with arrow keys from frozen to non-frozen which was scrolled such that the leftmost column needs to be scrolled into view

File: src/table/virtualizedDataTable.jsx [hunk 14]
```diff
@@ -1460,101 +1641,161 @@ class VirtualizedDataTable extends Component {
...
+            ))}
+          </Fragment>
+        ))}
         <KeyHandler
-          keys={keysToSniff}
+          keys={KEYS_TO_SNIFF}
           onKey={this.prvHandleKeyPress}
-          getInputRef={() => this.prvCellGrid}
+          getInputRef={() => [this.prvCellGridRef, this.prvCellGridFreezeRef]}
         />
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1690]: The following function was inlined below

File: src/table/virtualizedDataTable.jsx [hunk 15]
```diff
@@ -1460,101 +1641,161 @@ class VirtualizedDataTable extends Component {
...
+        ))}
         <div
-          style={_.assign({}, styles.headerDiv, {
+          style={{
+            ...STYLES.headerDiv,
             width,
             overflowY: noHeaderScroll ? 'hidden' : 'scroll',
-          })}
+          }}
         >
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1722]: Frozen part of header

File: src/table/virtualizedDataTable.jsx [hunk 16]
```diff
@@ -1460,101 +1641,161 @@ class VirtualizedDataTable extends Component {
...
           />
         </div>
         <div
-          style={_.assign({}, styles.tableDiv, {
+          style={{
+            ...STYLES.tableDiv,
             width,
-          })}
+          }}
         >
```
Comment (paulbrom) [inline src/table/virtualizedDataTable.jsx:1770]: Regular frozen columns

**Comment by paulbrom** [comment]
To build this and riun it in your browser:
```
npm install
npm run build:demo
```
then navigate to `virtualized-data-table/demo/index.html` (local file) in your browser

**Review by kanzelm3 — APPROVED** [review]
LGTM! 🎉

**Comment by craigmr** [comment]
How do you make a column freeze, it would be good to have some documentation in the Read me and or in the code.

**Comment by paulbrom** [comment]
> How do you make a column freeze, it would be good to have some documentation in the Read me and or in the code.

Yeah, the documentation for this package could use some love.   I will address that on a future 10% day...  but to answer your question, you add property **freezeLeftCount** to the <Table> component exposed by this package.  If this value is 3, then the 3 leftmost columns will be frozen.

**Pull request merged**